File: android/app/src/main/kotlin/io/opi/input/keyboard/SymbolPanel.kt
package io.opi.input.keyboard

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

enum class SymbolTab { Common, Emoji, All }

/**
 * 符号面板：常用/表情/全部 Tab + 关键字搜索（对齐 flutter symbol_panel.dart）。
 * 搜索输入焦点联动 qwerty 叠盘（IME 窗口内 TextField 不会唤起系统键盘）；
 * 防抖在 ImeState（250ms），本组件只按 searchQuery 出结果。
 */
@OptIn(ExperimentalFoundationApi::class)
@Composable
fun SymbolPanel(
    catalog: SymbolCatalog,
    searchText: String,
    searchActive: Boolean,
    searchQuery: String,
    onSearchText: (String) -> Unit,
    onSearchFocus: (Boolean) -> Unit,
    onCommit: (String) -> Unit,
    onClose: () -> Unit,
    onBackToNumber: () -> Unit,
) {
    var tab by remember { mutableStateOf(SymbolTab.Common) }

    fun commit(text: String) {
        if (SymbolCatalog.isEmoji(text)) catalog.recordRecent(text)
        onCommit(text)
    }

    Column(modifier = Modifier.fillMaxSize()) {
        // 头部：ABC / 123 + 搜索框
        Row(modifier = Modifier.fillMaxWidth().height(48.dp)) {
            KeyButton("ABC", modifier = Modifier.weight(2f), onTap = onClose)
            KeyButton("123", modifier = Modifier.weight(2f), onTap = onBackToNumber)
            Box(modifier = Modifier.weight(5f).padding(horizontal = 4.dp)) {
                TextField(
                    value = searchText,
                    onValueChange = onSearchText,
                    modifier = Modifier.fillMaxSize().onFocusChanged { onSearchFocus(it.isFocused) },
                    singleLine = true,
                    textStyle = TextStyle(fontSize = 14.sp),
                    placeholder = { Text("搜索（拼音/英文）", fontSize = 14.sp, color = Color(0xFF9E9E9E)) },
                    colors = TextFieldDefaults.colors(
                        focusedContainerColor = Color.White,
                        unfocusedContainerColor = Color.White,
                        focusedIndicatorColor = Color.Transparent,
                        unfocusedIndicatorColor = Color.Transparent,
                    ),
                )
            }
        }
        // Tab 栏
        Row(modifier = Modifier.fillMaxWidth()) {
            for (t in SymbolTab.entries) {
                Box(
                    modifier = Modifier
                        .weight(1f)
                        .height(36.dp)
                        .background(if (t == tab) Color(0xFFE0E0E0) else Color.Transparent)
                        .clickable { tab = t },
                    contentAlignment = Alignment.Center,
                ) {
                    Text(tabLabel(t), fontSize = 14.sp)
                }
            }
        }
        // 内容区：搜索激活且有关键字 → 结果网格（表情 Tab 过滤）；否则按 Tab 出内容
        Box(modifier = Modifier.weight(1f).fillMaxWidth()) {
            if (searchActive && searchQuery.isNotEmpty()) {
                val results = catalog.search(searchQuery)
                val shown = if (tab == SymbolTab.Emoji) results.filter(SymbolCatalog::isEmoji) else results
                SymbolGrid(shown, ::commit)
            } else {
                when (tab) {
                    SymbolTab.Common -> SymbolGrid(catalog.common, ::commit)
                    SymbolTab.Emoji -> Column(modifier = Modifier.fillMaxSize()) {
                        if (catalog.recents.isNotEmpty()) RecentsRow(catalog.recents, onCommit)
                        Box(modifier = Modifier.weight(1f).fillMaxWidth()) {
                            SymbolGrid(catalog.emoji, ::commit)
                        }
                    }
                    SymbolTab.All -> SymbolGrid(catalog.all, ::commit)
                }
            }
        }
    }
}

private fun tabLabel(t: SymbolTab): String = when (t) {
    SymbolTab.Common -> "常用"
    SymbolTab.Emoji -> "表情"
    SymbolTab.All -> "全部"
}

/** 最近使用表情行。 */
@Composable
private fun RecentsRow(recents: List<String>, onCommit: (String) -> Unit) {
    LazyRow(modifier = Modifier.fillMaxWidth().height(44.dp)) {
        itemsIndexed(recents) { _, text ->
            Box(
                modifier = Modifier.fillMaxSize().clickable { onCommit(text) },
                contentAlignment = Alignment.Center,
            ) {
                Text(text, fontSize = 22.sp, maxLines = 1, overflow = TextOverflow.Ellipsis)
            }
        }
    }
}

/** 8 列符号网格（惰性构建：全量条目数千，count 构建不展开）。 */
@OptIn(ExperimentalFoundationApi::class)
@Composable
private fun SymbolGrid(entries: List<String>, onTap: (String) -> Unit) {
    LazyVerticalGrid(columns = GridCells.Fixed(8), modifier = Modifier.fillMaxSize()) {
        items(entries.size) { i ->
            val text = entries[i]
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(2.dp)
                    .clip(RoundedCornerShape(6.dp))
                    .clickable { onTap(text) },
                contentAlignment = Alignment.Center,
            ) {
                Text(text, fontSize = 20.sp, maxLines = 1, overflow = TextOverflow.Ellipsis)
            }
        }
    }
}

File: flutter/app/lib/keyboards/symbol_panel.dart
import 'dart:async';

import 'package:app/keyboards/key_button.dart';
import 'package:app/keyboards/symbol_catalog.dart';
import 'package:app/src/rust/api.dart';
import 'package:flutter/material.dart';

enum SymbolTab { common, emoji, all }

/// 符号面板：常用/表情/全部 Tab + 关键字搜索。
/// 搜索输入焦点联动 qwerty（IME 窗口内 TextField 不会唤起系统键盘），
/// 输入框 controller 由父级持有；此处监听其变化并 250ms 防抖出结果。
class SymbolPanel extends StatefulWidget {
  const SymbolPanel({
    super.key,
    required this.catalog,
    required this.onCommit,
    required this.onClose,
    required this.onBackToNumber,
    required this.searchController,
    required this.searchFocusNode,
    required this.searchActive,
  });

  final SymbolCatalog catalog;
  final ValueChanged<String> onCommit;
  final VoidCallback onClose;
  final VoidCallback onBackToNumber;
  final TextEditingController searchController;
  final FocusNode searchFocusNode;

  /// 焦点态：true 时下方显示 qwerty 字母盘供搜索输入。
  final bool searchActive;

  @override
  State<SymbolPanel> createState() => _SymbolPanelState();
}

class _SymbolPanelState extends State<SymbolPanel> {
  SymbolTab _tab = SymbolTab.common;
  Timer? _debounce;
  String _query = '';

  @override
  void initState() {
    super.initState();
    widget.searchController.addListener(_onInputChanged);
  }

  @override
  void dispose() {
    _debounce?.cancel();
    widget.searchController.removeListener(_onInputChanged);
    super.dispose();
  }

  /// 统一入口：TextField 键入与 qwerty 搜索盘程序化修改都经 controller 监听。
  void _onInputChanged() {
    _debounce?.cancel();
    _debounce = Timer(const Duration(milliseconds: 250), () {
      if (mounted) setState(() => _query = widget.searchController.text.trim());
    });
  }

  void _commit(ApiSymbolEntry entry) {
    if (entry.emoji) widget.catalog.recordRecent(entry.text);
    widget.onCommit(entry.text);
  }

  String _tabLabel(SymbolTab t) => switch (t) {
        SymbolTab.common => '常用',
        SymbolTab.emoji => '表情',
        SymbolTab.all => '全部',
      };

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        _header(),
        _tabBar(),
        Expanded(child: _body()),
      ],
    );
  }

  Widget _header() {
    return Row(
      children: [
        KeyButton('ABC', flex: 2, onTap: widget.onClose),
        KeyButton('123', flex: 2, onTap: widget.onBackToNumber),
        Expanded(
          child: Padding(
            padding: const EdgeInsets.symmetric(horizontal: 4),
            child: TextField(
              controller: widget.searchController,
              focusNode: widget.searchFocusNode,
              textInputAction: TextInputAction.search,
              decoration: const InputDecoration(
                hintText: '搜索（拼音/英文）',
                isDense: true,
                border: OutlineInputBorder(),
                contentPadding: EdgeInsets.symmetric(horizontal: 8),
              ),
            ),
          ),
        ),
      ],
    );
  }

  Widget _tabBar() {
    return Row(
      children: [
        for (final tab in SymbolTab.values)
          Expanded(
            child: InkWell(
              onTap: () => setState(() => _tab = tab),
              child: Container(
                height: 36,
                color: _tab == tab ? Colors.grey.shade300 : Colors.transparent,
                alignment: Alignment.center,
                child: Text(_tabLabel(tab)),
              ),
            ),
          ),
      ],
    );
  }

  Widget _body() {
    if (widget.searchActive && _query.isNotEmpty) {
      final results = widget.catalog.search(_query);
      final shown =
          _tab == SymbolTab.emoji ? results.where((e) => e.emoji).toList() : results;
      return _grid(shown);
    }
    switch (_tab) {
      case SymbolTab.common:
        return _grid(widget.catalog.common);
      case SymbolTab.emoji:
        return Column(
          children: [
            if (widget.catalog.recents.isNotEmpty) _recentsRow(),
            Expanded(child: _grid(widget.catalog.emoji)),
          ],
        );
      case SymbolTab.all:
        return _grid(widget.catalog.all);
    }
  }

  Widget _recentsRow() {
    final recents = widget.catalog.recents;
    return SizedBox(
      height: 44,
      child: ListView(
        scrollDirection: Axis.horizontal,
        children: [
          for (final text in recents)
            InkWell(
              onTap: () => widget.onCommit(text),
              child: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 10),
                child: Center(child: Text(text, style: const TextStyle(fontSize: 22))),
              ),
            ),
        ],
      ),
    );
  }

  /// GridView.builder 惰性构建（全量条目数千，count 列表会全量构建）。
  Widget _grid(List<ApiSymbolEntry> entries) {
    return GridView.builder(
      gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(crossAxisCount: 8),
      itemCount: entries.length,
      itemBuilder: (context, i) {
        final entry = entries[i];
        return InkWell(
          onTap: () => _commit(entry),
          child: Center(
            child: Text(
              entry.text,
              maxLines: 1,
              overflow: TextOverflow.ellipsis,
              style: const TextStyle(fontSize: 20),
            ),
          ),
        );
      },
    );
  }
}
